### PR TITLE
feat(parser): the 1st version of a new combinator style parser

### PR DESCRIPTION
**Pull request by TennyZhuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Introducing a new parser with [winnow](https://github.com/winnow-rs/winnow).

This is the first version of the new parser, with only the datatype parsing. However, it's also the most complicated part of our Parser. Due to our wrong design previously, we use `<` and `>` to define struct type, then introduce a necessary state `angle_number` during parsing. Fortunately, the new framework handle it perfectly.

winnow-rs is a parser-combinator crate, and it's forked by me for a simple patch https://github.com/winnow-rs/winnow/pull/525.

**The new framework can progressively replace the original parser, see `Parser::parse_v2` for detail.**

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/sqlparser/src/parser_v2/data_type.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        Keyword::STRING | Keyword::VARCHAR => empty.value(DataType::Varchar),
+        Keyword::CHAR | Keyword::CHARACTER => dispatch! {keyword;
+            Keyword::VARYING => empty.value(DataType::Varchar),
+            _ => opt(precision_in_range(..)).map(|precision| DataType::Char(precision)),
+        },
+        Keyword::UUID => empty.value(DataType::Uuid),
+        Keyword::DATE => empty.value(DataType::Date),
+        Keyword::TIMESTAMP => with_time_zone().map(|with_tz| DataType::Timestamp(with_tz)),
+        Keyword::TIME => with_time_zone().map(|with_tz| DataType::Time(with_tz)),
+        // TODO: Support complex inverval type parsing.
```
Comment (TennyZhuang) [inline src/sqlparser/src/parser_v2/data_type.rs:138]:
```suggestion
        // TODO: Support complex interval type parsing.
```

**Comment by wangrunji0408** [comment]
Can we have a better error display with this new parser framework? 🤔



taken from #15091

**Comment by TennyZhuang** [comment]
> Can we have a better error display with this new parser framework? if it provides span in some way. 🤔

It's possible, but must after we finish the whole refactor.

File: src/sqlparser/src/parser_v2/data_type.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+where
+    S: TokenStream,
+{
+    #[derive(Debug, thiserror::Error)]
+    #[error("Unconsumed `>>`")]
+    struct UnconsumedShiftRight;
+
+    with_state::<S, DataTypeParsingState, _, _>(terminated(
+        data_type_stateful,
+        cut_err(trace(
```
Comment (BugenZhao) [inline src/sqlparser/src/parser_v2/data_type.rs:119]: `cut_err` here seems unnecessary as the inner parser always returns `ErrMode::Cut`.

File: src/sqlparser/src/parser_v2/data_type.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        Keyword::INTERVAL => empty.value(DataType::Interval),
+        Keyword::REGCLASS => empty.value(DataType::Regclass),
+        Keyword::REGPROC => empty.value(DataType::Regproc),
+        Keyword::TEXT => empty.value(DataType::Text),
+        Keyword::STRUCT => cut_err(struct_data_type).map(DataType::Struct),
+        Keyword::BYTEA => empty.value(DataType::Bytea),
+        Keyword::NUMERIC | Keyword::DECIMAL | Keyword::DEC => cut_err(precision_and_scale()).map(|(precision, scale)| {
+            DataType::Decimal(precision, scale)
+        }),
+        _ => fail,
```
Comment (BugenZhao) [inline src/sqlparser/src/parser_v2/data_type.rs:204]: Hitting this branch means that we've successfully got a keyword but it does not match any data type. Should we `cut_err` here?

File: src/sqlparser/src/parser.rs [hunk 4]
```diff
@@ -172,19 +173,51 @@ pub struct Parser {
...
+            };
+            ParserError::ParserError(format!(
+                "Unexpected {}{}",
+                if self.index + token_stream.location() >= self.tokens.len() {
+                    &"EOF" as &dyn std::fmt::Display
+                } else {
+                    &self.tokens[self.index + token_stream.location()] as &dyn std::fmt::Display
+                },
+                msg
+            ))
```
Comment (BugenZhao) [inline src/sqlparser/src/parser.rs:216]: Can we use `ParseError` here for better error reporting?

**Review by BugenZhao — COMMENTED** [review]
The attempt in this PR is truly exciting. The combinator style makes the code more readable and maintainable (although `cut_err` could be a bit subtle 🤯). Looking forward to seeing the user-facing benefits (like better error reporting) after the refactoring.

File: src/sqlparser/src/parser_v2/mod.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+/// Bundle trait requirements from winnow, so that we don't need to write them everywhere.
+///
+/// All combinators should accept a generic `S` that implements `TokenStream`.
+pub trait TokenStream:
+    Stream<Token = TokenWithLocation> + StreamIsPartial + Location + Default
+{
+}
+
+impl<S> TokenStream for S where
```
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/sqlparser/src/parser_v2/mod.rs:39]: rust-analyzer doesn't has good support for this feature. I just tried it and the development experience is poor.

File: src/sqlparser/src/parser_v2/mod.rs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+where
+    S: TokenStream,
+{
+    struct WhiteSpace;
+
+    impl ContainsToken<TokenWithLocation> for WhiteSpace {
+        fn contains_token(&self, token: TokenWithLocation) -> bool {
+            matches!(token.token, Token::Whitespace(_))
+        }
+    }
```
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/sqlparser/src/parser_v2/mod.rs:65]: image
Just found it, interesting

File: src/sqlparser/src/parser_v2/impl_.rs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    #[inline(always)]
+    fn reset(&mut self, checkpoint: &Self::Checkpoint) {
+        self.tokens.reset(&checkpoint.0)
+    }
+
+    #[inline(always)]
+    fn raw(&self) -> &dyn std::fmt::Debug {
+        self
```
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/sqlparser/src/parser_v2/impl_.rs:141]: No, customizing `Debug` is the main purpose of the wrapper.

File: src/sqlparser/src/parser.rs (hunk 4 above)
Comment (TennyZhuang) [reply src/sqlparser/src/parser.rs:216]: Which `ParserError`? If you mean v2, I'd prefer give user a consistent user experience now.

File: src/sqlparser/src/parser_v2/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/sqlparser/src/parser_v2/mod.rs:39]: Makes sense. It will become `impl Sized`. 😁

**Comment by xxchan** [comment]
> The combinator style makes the code more readable and maintainable

Why do you think the combinator is more maintainable? I think now everyone can work with the parser without learning, but not afterwards.

**Comment by TennyZhuang** [comment]
> Why do you think the combinator is more maintainable? I think now everyone can work with the parser without learning, but not afterwards.

Let's use Golang to rewrite RisingWave 😄

**Review by xxchan — APPROVED** [review]
rubber stamp

**Pull request merged**